grommet/hpe-design-system
# Guidance Update Feature

[Deploy Preview](https://deploy-preview-3502--keen-mayer-a86c8b.netlify.app/)

#### What does this PR do?
Added an update feature that tracks changes to the guidances. It uses the Github API to read the pull requests closed onto the site, then displays the information via inline notifications and tags. It also uses visitation tracking for a personalized experience, this makes it so once the user has seen the update on the guidance page, it won't show again. 

#### Where should the reviewer start?
They should start in the app.js file in the pages folder. This has the fetching information and calls the other components with that information. It would then be beneficial to explore the functions called/added in app.js from there. 

#### What testing has been done on this PR?
I've tested that the feature against older PRs (making sure they wrote over the old ones), tested it in incognito with no access to local storage, and different cases with last-visted tokens.

In addition to the feature you are implementing, have you checked the following:

**General UX Checks**

**Accessibility Checks**

**Code Quality Checks**

#### How should this be manually tested?
To test visitation tracking or to see how information is being stored, I would check the local storage in the browser's developer tools. A user who is entirely new to the site will have neither "update-history" or any tokens ending in "-last-visited," but when launching, it will create the "update-history."

#### Any background context you want to provide?
To run the visitation tracking correctly (in dev mode) I had to turn the React StrictMode to false when writing it. I turned it back to true before this push/pull, but to run it with proper logic it should be toggled to false.

#### What are the relevant issues?
Issue #3405 

#### Screenshots (if appropriate)
![Screenshot 2023-08-07 at 11 03 16 AM](https://github.com/grommet/hpe-design-system/assets/123425454/cd3d48a1-9477-4eee-8650-a4b899e74b0b)

![Screenshot 2023-08-07 at 11 02 51 AM](https://github.com/grommet/hpe-design-system/assets/123425454/4ab886cc-ad55-486b-ab5f-82a6519927cc)

#### Should this PR be mentioned in Design System updates?
Yes

#### Is this change backwards compatible or is it a breaking change?

#### Note about pageUpdateReady:
In app.js there is a declared state of pageUpdateReady. This is used to accommodate for the rendering behavior when entering a guidance page. Here is the context we need it for:
- A user who hasn't seen the update enters the guidance page, sees the update, and navigates back to the components page.
- The update state for this page will still be true as they navigate away, this is because the update state is only calculated when they enter the guidance page.
- The user clicks on the component card to re-enter the guidance page they just saw. This is how it renders:

1. Layout.js renders the current update state (which would be true before calculations)
2. The route change hook in app.js sets the update state as false
3. Layout.js re-renders the page with the update as false (this switch from true -> false causes the flash)

- We address this with the pageUpdateReady state variable by changing the flow to this:
1. Layout.js renders, but it first (and always) sets the pageUpdateReady as false. Since the update components' rendering rely on this to be true, they won't show yet.
2. The route change hook in app.js then sets the update state as false (the correct state.) Once this is figured out, it will set pageUpdateReady as true.
3. Since the pageUpdateReady is now true, Layout.js re-renders with pageUpdateReady as true (which allows for the inline notifications/side navigation indicators to load the proper information once.)
